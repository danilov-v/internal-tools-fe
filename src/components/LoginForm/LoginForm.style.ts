import styled from 'styled-components';

type InputProps = {
  isValid?: boolean;
  marginBottom?: number;
};

export const Form = styled.form`
  text-align: center;
`;

export const Input = styled.input`
  background-color: transparent;
  border: 0;
  border-bottom: ${(props: InputProps) =>
    `2px solid ${props.isValid ? '#ccccccff' : '#ff0000'}`};
  color: #ccccccff;
  font-size: 22px;
  margin-bottom: ${(props: InputProps) => `${props.marginBottom || 48}px`};
  outline: none;
  text-align: center;
  font-weight: 300;
  padding: 8px 0;
`;

export const SubmitButton = styled.button`
  color: #ccccccff;
  background-color: transparent;
  padding: 12px;
  border-radius: 4px;
  border: 1px solid #ccccccff;
  text-transform: uppercase;
  font-weight: 300;
  font-size: 18px;
  cursor: pointer;
  &:hover {
    opacity: 0.8;
  }
`;
