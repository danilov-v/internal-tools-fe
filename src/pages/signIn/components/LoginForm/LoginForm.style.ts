import styled from 'styled-components';

type InputProps = {
  isValid?: boolean;
  marginBottom?: number;
};

export const LoginForm = styled.form`
  text-align: center;
`;

export const InputField = styled.input<InputProps>`
  background-color: transparent;
  border: 0;
  border-bottom: ${props =>
    `2px solid ${props.isValid ? '#cccccc' : '#ff0000'}`};
  color: #cccccc;
  font-size: 22px;
  margin-bottom: ${props => `${props.marginBottom || 48}px`};
  outline: none;
  text-align: center;
  font-weight: 300;
  padding: 8px 0;
`;

export const SubmitButton = styled.button`
  color: #cccccc;
  background-color: transparent;
  padding: 12px;
  border-radius: 4px;
  border: 1px solid #cccccc;
  text-transform: uppercase;
  font-weight: 300;
  font-size: 18px;
  cursor: pointer;
  &:hover {
    opacity: 0.8;
  }
`;
