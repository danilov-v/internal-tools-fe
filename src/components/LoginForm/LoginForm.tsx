import React, { useState } from 'react';
import { useNavigate } from '@reach/router';
import { checkAuth } from 'services/auth';

import * as S from './LoginForm.style';

type FormData = {
  login: string;
  password: string;
};

const DEFAULT_FORM_DATA = {
  login: '',
  password: '',
};

export const LoginForm: React.FC<{}> = () => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState<FormData>(DEFAULT_FORM_DATA);
  const [isValid, setValid] = useState<boolean>(true);

  const handleInput = (field: string) => (
    e: React.FormEvent<HTMLInputElement>,
  ): void => {
    e.preventDefault();

    setValid(true);
    setFormData({
      ...formData,
      [field]: e.currentTarget.value,
    });
  };

  const handleLoginInput = handleInput('login');
  const handlePasswordInput = handleInput('password');

  const resetForm = (): void => {
    setFormData(DEFAULT_FORM_DATA);
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    if (checkAuth(formData)) {
      navigate('/main');
    } else {
      setValid(false);
      resetForm();
    }
  };

  return (
    <S.Form onSubmit={handleSubmit} data-testid="loginForm">
      <S.Input
        type="text"
        value={formData.login}
        placeholder="Капитан"
        onChange={handleLoginInput}
        isValid={isValid}
      />
      <br />
      <S.Input
        type="password"
        value={formData.password}
        placeholder="Пароль"
        onChange={handlePasswordInput}
        isValid={isValid}
        marginBottom={80}
      />
      <br />
      <S.SubmitButton type="submit" name="submit">
        Войти в ИТ
      </S.SubmitButton>
    </S.Form>
  );
};
