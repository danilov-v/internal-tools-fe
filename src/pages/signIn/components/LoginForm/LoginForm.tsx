import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from '@reach/router';

import { requestLogin } from 'redux/profile/thunks';
import { isAuthorizedProfile, getProfileError } from 'redux/profile/selectors';

import * as S from './LoginForm.style';

const DEFAULT_FORM_DATA = {
  login: '',
  password: '',
};

export const LoginForm: React.FC = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const [formData, setFormData] = useState(DEFAULT_FORM_DATA);
  const [isPristine, setPristine] = useState<boolean>(true);

  const isAuthorized = useSelector(isAuthorizedProfile);
  const profileError = useSelector(getProfileError);

  const isValid = isPristine || !profileError;

  const handleInput = (field: 'login' | 'password') => (
    e: React.FormEvent<HTMLInputElement>,
  ): void => {
    e.preventDefault();

    setFormData({ ...formData, [field]: e.currentTarget.value });
  };

  useEffect(() => {
    const resetPassword = (): void => {
      setFormData(state => ({ ...state, password: '' }));
    };

    if (!isValid) {
      resetPassword();
    }
  }, [isValid]);

  const handleLoginInput = handleInput('login');
  const handlePasswordInput = handleInput('password');

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    setPristine(false);

    dispatch(requestLogin(formData));
  };

  if (isAuthorized) {
    navigate('/personnel');
    return null;
  }

  return (
    <S.LoginForm onSubmit={handleSubmit} data-testid="loginForm">
      <S.InputField
        type="text"
        value={formData.login}
        placeholder="Имя пользователя"
        onChange={handleLoginInput}
        isValid={isValid}
      />
      <br />
      <S.InputField
        type="password"
        value={formData.password}
        placeholder="Пароль"
        onChange={handlePasswordInput}
        isValid={isValid}
        marginBottom={80}
      />
      <br />
      <S.SubmitButton type="submit" name="submit">
        Войти в ИТ
      </S.SubmitButton>
    </S.LoginForm>
  );
};
