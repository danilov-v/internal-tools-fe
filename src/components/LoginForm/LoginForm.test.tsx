import React from 'react';
import { render } from '@testing-library/react';

import { LoginForm } from './LoginForm';

describe('<LoginForm />', () => {
  it('should render and match snapshot', () => {
    const component = render(<LoginForm />);

    expect(component.baseElement).toMatchSnapshot();
  });
});
