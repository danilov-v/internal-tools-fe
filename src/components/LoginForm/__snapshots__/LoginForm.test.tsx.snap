// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should render and match snapshot 1`] = `
<body>
  <div>
    <form
      class="sc-AxjAm jWlVxK"
      data-testid="loginForm"
    >
      <input
        class="sc-AxirZ gQaiPr"
        placeholder="Капитан"
        type="text"
        value=""
      />
      <br />
      <input
        class="sc-AxirZ dsuoSr"
        placeholder="Пароль"
        type="password"
        value=""
      />
      <br />
      <button
        class="sc-AxiKw fFlbgl"
        name="submit"
        type="submit"
      >
        Войти в ИТ
      </button>
    </form>
  </div>
</body>
`;
